### ISSUE TITLE
Migration block not calling in multiple realm files

### ISSUE THREAD

**Issue by jplazcano87** [realm/realm-cocoa#2640]
i need to run a migration in multiple .realm files, i do this, but the migration block is never called, the app continues running and crash due the lack of migration i put a breakpoint and found out that the migration block is never been called

```
//migration block
RLMMigrationBlock migrationBlock =  ^(RLMMigration *migration, uint64_t oldSchemaVersion) {
    
    if (oldSchemaVersion < 1) {
      [migration enumerateObjects:ScreenModel.className block:^(RLMObject *oldObject, RLMObject *newObject) {
        newObject[@"was_showed"] = @NO;
        
      }];
    }
  };

  NSMutableArray *pathArray = [[NSMutableArray alloc] init];
  NSArray *dirpaths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
  NSString *docsDir =dirpaths[0];
  
  
  NSArray* dirs = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:docsDir
                                                                      error:NULL];
  ///enumeration of .realm files in the filesystem
  [dirs enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
    NSString *filename = (NSString *)obj;
    NSString *extension = [[filename pathExtension] lowercaseString];
    if ([extension isEqualToString:@"realm"]) {
      
      NSString *defaultRealmPath = [RLMRealmConfiguration defaultConfiguration].path;
      NSString *defaultRealmParentPath = [defaultRealmPath stringByDeletingLastPathComponent];
 
      NSLog(@"path %@",[defaultRealmParentPath stringByAppendingPathComponent:filename]);
// run the migration for each .ream file in the filesystem
      RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
      config.schemaVersion = 1;
      config.migrationBlock = migrationBlock;
      config.path = [defaultRealmParentPath stringByAppendingPathComponent:filename];
      [RLMRealm migrateRealm:config];
    }
  }];
```

**Comment by jplazcano87** [comment]
i have one .realm file per user logged into the app so i need the [dirs enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop)  to find out how many realm files i have, also the realm files are created with a uuid from a webservice so, i can't know the . realm name for sure

**Comment by kishikawakatsumi** [comment]
@jplazcano87  `+[RLMRealm migrateRealm:]` method is called automatically when opening a Realm for the first time and does not need to be called explicitly.

So you should pass `RLMRealmConfiguration` object with migration block to the `+[RLMRealm realmWithConfiguration:error:]` method when you open the Realm.

If you would like to execute migration process explicitly, you can change above code to like below:

```
[dirs enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
    ...
    RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
    config.schemaVersion = 1;
    config.migrationBlock = migrationBlock;
    config.path = [defaultRealmParentPath stringByAppendingPathComponent:filename];
    
    NSError *error = nil;
    [RLMRealm realmWithConfiguration:config error:&error];
```

**Comment by jplazcano87** [comment]
there is any other way to migrate multiple realm files with the same block ? 

```
  [dirs enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
    NSString *filename = (NSString *)obj;
    NSString *extension = [[filename pathExtension] lowercaseString];
    if ([extension isEqualToString:@"realm"]) {
  
      RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
      config.schemaVersion = 1;
      config.migrationBlock = block;
      config.path =[defaultRealmParentPath stringByAppendingPathComponent:filename];
      NSError *error = nil;

      [RLMRealm realmWithConfiguration:config error:&error];
    
      
    }
  }];

```

i just do this, and now am getting this error 'Provided schema version 0 is less than last set version 1.'
do you have any recommendation to migrate multiple .realm files?

thanks in advice, and sorry for the bad english, i'm from Chile

**Comment by daniel-beard** [comment]
I'm seeing the same problem: When calling `+[RLMRealm realmWithConfiguration:error:]` (and I have multiple realms within the same app).

I get the following from the error param:

```
objc[86808]: EXCEPTIONS: catch(NSException)
objc[86808]: EXCEPTIONS: unwinding through frame [ip=0x10d5c3e46 sp=0x7fff536c9240] for exception 0x7fabc0240b18
objc[86808]: EXCEPTIONS: handling exception 0x7fabc0240b18 at 0x10d5c3f91
objc[86808]: EXCEPTIONS: finishing handler
objc[86808]: EXCEPTIONS: releasing completed exception 0x7fabc0240b38 (object 0x7fabc0240e80, a NSException)
Printing description of error:
Error Domain=io.realm Code=0 "Migration is required for object type 'MyObject' due to the following errors:
- Property 'primaryKey' has been made optional.
- Property '[REDACTED]' has been made optional.
- Property 'today' has been made optional." UserInfo={NSLocalizedDescription=Migration is required for object type 'MyObject' due to the following errors:
- Property 'primaryKey' has been made optional.
- Property '[REDACTED]' has been made optional.
- Property 'today' has been made optional.}
```

**Comment by jplazcano87** [comment]
@daniel-beard did you find any way to fix it?

**Comment by daniel-beard** [comment]
@jplazcano87 can't figure out how to fix it, but I did put together a sample project that shows the problem: https://github.com/daniel-beard/RealmTest/tree/master

**Comment by bdash** [comment]
@daniel-beard, that's a great test app. From a brief look I can see two issues, one in Realm and one in the test code:

1. `-[RLMRealm migrateRealm:]` doesn't appear to respect the schema version from the configuration, instead using only the schema version set via the legacy APIs. @tgoyne's change in 16194870 fixes this issue. We should pull that change into a point update.
2. With version 0.96 you'll need to explicitly state that your `NSString *`, `NSDate *` and `NSData *` properties are required by implementing `+requiredProperties` in each model class to avoid requiring a migration. Those data types are now considered to be nullable unless `+requiredProperties` says otherwise.

**Comment by daniel-beard** [comment]
Thanks @bdash for the quick turn-around!

**Comment by jpsim** [comment]
Looks like @bdash and I both looked at this at the same time. We mostly agree, but I think both "bugs" are really ours.

---

Thank you so much for sharing that, it's really clarified what's happening. There are two bugs here. One of them is in our code, the other is in our documentation ;)

The bug in our code is that we're using the schema version set using [RLMRealm setSchemaVersion:...] rather than the one provided in the configuration object. We've already submitted a fix for this and it will land in our next release: https://github.com/realm/realm-cocoa/pull/2680

The bug in our docs is that the following sentence is wrong:

> Accessing a Realm file created with a previous version will automatically convert these properties to nullable in the file itself, unless explicitly marked not to do so in +requiredProperties.

Although accessing the Realm file will in fact update the file format, the properties won't be converted to nullable until you perform a migration. We'll be sure to update the 0.96 release blog post and our docs to reflect this.

Thanks again for sending this our way.

**Comment by daniel-beard** [comment]
Thanks @jpsim, I was sure I was setting something up incorrectly :D

**Comment by bdash** [comment]
The bug that resulted in the schema version on the configuration being ignored was fixed via #2707, which was released in v0.96.1.
